File: packages/node/src/integrate.ts
/** `brainpick integrate <target>`: meet agents where they live. Ports integrate.py.
 *
 * Three targets, one family voice (mirrors henxels' `integrate`):
 *
 * - `claude-code`  — write the Agent Skill into the repo, then PRINT a paste-able
 *   graph-before-grep PreToolUse hook and the `claude mcp add` snippet (settings.json
 *   is never edited for you).
 * - `opencode`     — write the skill under OpenCode's convention, then PRINT the
 *   opencode.json MCP snippet.
 * - `agents-md`    — ensure an AGENTS.md exists (the one place integrate may create a
 *   file), install the brain-report markers if absent, and compile so the block fills.
 *
 * The shipped Agent Skill (integrations/skill/SKILL.md, canonical) rides inside the
 * package; the parity test asserts the shipped copy is byte-identical to the canonical.
 */
import { existsSync, mkdirSync, readFileSync, statSync, writeFileSync } from "node:fs";
import { dirname, join, resolve } from "node:path";

import { runCompile } from "./compile/pipeline";
import { REPORT_BEGIN_PREFIX, REPORT_END_MARKER } from "./compile/t1";
import { findRepoRoot } from "./detect";
import { brainpickCommand, mcpSnippets, Voice, type Print } from "./scaffold";
import { PACKAGE_ROOT } from "./version";

export const TARGETS = ["claude-code", "opencode", "agents-md"] as const;
export type Target = (typeof TARGETS)[number];

const HENXELS_BEGIN = "<!-- henxels:begin -->";

// harness -> where its Agent Skill lands, relative to the repo root
export const SKILL_DESTINATIONS: Record<"claude-code" | "opencode", string> = {
  "claude-code": join(".claude", "skills", "brainpick", "SKILL.md"),
  opencode: join(".opencode", "skills", "brainpick", "SKILL.md"),
};

const MINIMAL_AGENTS = "# AGENTS.md\n\nWorking notes for agents in this repository.\n";
const REPORT_PLACEHOLDER =
  `${REPORT_BEGIN_PREFIX}pending) -->\n` +
  "_brainpick fills this block on the next `brainpick compile`._\n" +
  REPORT_END_MARKER;

/** The shipped Agent Skill: the package copy first (installed tarballs), then the
 * repo-root canonical (dev checkout). */
export function skillPath(): string {
  const packaged = resolve(PACKAGE_ROOT, "skill", "SKILL.md");
  try {
    if (statSync(packaged).isFile()) return packaged;
  } catch {
    /* installed without the shipped copy — fall through to the repo canonical */
  }
  return resolve(PACKAGE_ROOT, "..", "..", "integrations", "skill", "SKILL.md");
}

export function skillText(): string {
  return readFileSync(skillPath(), "utf8");
}

/** Python's shlex.quote: bare when every character is safe, else single-quoted. */
function shellQuote(word: string): string {
  if (word === "") return "''";
  return /[^\w@%+=:,./-]/.test(word) ? `'${word.replaceAll("'", "'\"'\"'")}'` : word;
}

/** A paste-able Claude Code hooks fragment: a PreToolUse nudge toward the brain before
 * it greps — advisory (exit 0), never a block — and the UserPromptSubmit recall that puts
 * the matching memories into context (spec/72). */
function hooksFragment(root: string): string {
  const recall = [...brainpickCommand(), "recall", "--root", root].map(shellQuote).join(" ");
  const fragment = {
    hooks: {
      PreToolUse: [
        {
          matcher: "Grep|Glob",
          hooks: [
            {
              type: "command",
              command:
                "echo 'brainpick: consult the brain first — brain_search " +
                "or `brainpick search` before grepping' >&2",
            },
          ],
        },
      ],
      UserPromptSubmit: [{ hooks: [{ type: "command", command: recall, timeout: 15 }] }],
    },
  };
  return JSON.stringify(fragment, null, 2);
}

function writeSkill(repo: string, target: "claude-code" | "opencode", dryRun: boolean): [string, boolean] {
  const dest = join(repo, SKILL_DESTINATIONS[target]);
  const existed = existsSync(dest);
  if (!dryRun) {
    mkdirSync(dirname(dest), { recursive: true });
    writeFileSync(dest, skillText(), "utf8");
  }
  return [dest, existed];
}

/** Install the report markers above the henxels digest block when one exists,
 * else at the end — never disturbing hand-written content. */
export function insertReportMarkers(text: string): string {
  if (text.includes(REPORT_BEGIN_PREFIX)) return text;
  const idx = text.indexOf(HENXELS_BEGIN);
  if (idx !== -1) {
    return text.slice(0, idx).replace(/\n+$/, "") + "\n\n" + REPORT_PLACEHOLDER + "\n\n" + text.slice(idx);
  }
  return text.replace(/\n+$/, "") + "\n\n" + REPORT_PLACEHOLDER + "\n";
}

function integrateClaudeCode(voice: Voice, root: string, repo: string, dryRun: boolean): number {
  const [dest, existed] = writeSkill(repo, "claude-code", dryRun);
  const verb = dryRun ? "would write" : existed ? "updated" : "wrote";
  voice.line("✓", `skill: ${verb} ${dest}`);
  if (dryRun) {
    voice.step("• print the graph-before-grep and recall hooks and the `claude mcp add` snippet");
    return 0;
  }
  voice.raw();
  voice.raw("Paste into .claude/settings.json (settings are never edited for you):");
  voice.raw();
  voice.raw(hooksFragment(root));
  voice.raw();
  voice.raw(mcpSnippets(root));
  return 0;
}

function integrateOpencode(voice: Voice, root: string, repo: string, dryRun: boolean): number {
  const [dest, existed] = writeSkill(repo, "opencode", dryRun);
  const verb = dryRun ? "would write" : existed ? "updated" : "wrote";
  voice.line("✓", `skill: ${verb} ${dest}`);
  if (dryRun) {
    voice.step("• print the opencode.json MCP snippet");
    return 0;
  }
  voice.raw();
  voice.raw("Add the MCP server to opencode.json (merging JSON is left to you):");
  voice.raw();
  voice.raw(mcpSnippets(root));
  return 0;
}

async function integrateAgentsMd(voice: Voice, root: string, repo: string, dryRun: boolean): Promise<number> {
  const agents = join(repo, "AGENTS.md");
  const existed = existsSync(agents);
  const hasMarkers = existed && readFileSync(agents, "utf8").includes(REPORT_BEGIN_PREFIX);

  if (dryRun) {
    if (!existed) voice.step(`• create a minimal ${agents}`);
    if (!hasMarkers) voice.step(`• install the brain-report markers in ${agents}`);
    voice.step(`• compile ${root} so the report block fills`);
    return 0;
  }

  let text = existed ? readFileSync(agents, "utf8") : MINIMAL_AGENTS;
  if (!existed) voice.line("✓", `AGENTS.md: created ${agents}`);
  if (!text.includes(REPORT_BEGIN_PREFIX)) {
    text = insertReportMarkers(text);
    voice.line("✓", `report: markers installed in ${agents}`);
  } else {
    voice.line("○", `report: markers already in ${agents}`);
  }
  writeFileSync(agents, text, "utf8");

  const result = await runCompile(root);
  voice.line("✓", `compiled: the report block is filled (seq ${result.seq})`);
  voice.step(`read it back: sed -n '/brainpick:begin report/,/brainpick:end report/p' ${agents}`);
  return 0;
}

export interface IntegrateOptions {
  dryRun?: boolean;
  env?: NodeJS.ProcessEnv;
  print?: Print;
}

export async function runIntegrate(target: string, root: string, options: IntegrateOptions = {}): Promise<number> {
  const print = options.print ?? ((line: string) => console.log(line));
  if (!(TARGETS as readonly string[]).includes(target)) {
    print(`unknown target '${target}'; choose from ${TARGETS.join(", ")}`);
    return 1;
  }
  const voice = new Voice(options.env ?? process.env, print);
  voice.banner();
  const resolved = resolve(root);
  let isDir = false;
  try {
    isDir = statSync(resolved).isDirectory();
  } catch {
    isDir = false;
  }
  if (!isDir) {
    voice.line("✗", `${resolved} is not a directory`);
    return 1;
  }
  const repo = findRepoRoot(resolved) ?? resolved;
  voice.line("○", `repo root: ${repo}` + (repo !== resolved ? "" : " (the bundle is its own repo)"));
  if (options.dryRun) {
    voice.raw();
    voice.raw(`dry run — nothing written. integrate ${target} would:`);
  }

  if (target === "claude-code") return integrateClaudeCode(voice, resolved, repo, Boolean(options.dryRun));
  if (target === "opencode") return integrateOpencode(voice, resolved, repo, Boolean(options.dryRun));
  return integrateAgentsMd(voice, resolved, repo, Boolean(options.dryRun));
}

File: packages/python/src/brainpick/integrate.py
"""`brainpick integrate <target>`: meet agents where they live.

Three targets, one family voice (mirrors henxels' `integrate`):

- `claude-code`  — write the Agent Skill into the repo, then PRINT a paste-able
  graph-before-grep PreToolUse hook and the `claude mcp add` snippet (settings.json
  is never edited for you).
- `opencode`     — write the skill under OpenCode's convention, then PRINT the
  opencode.json MCP snippet.
- `agents-md`    — ensure an AGENTS.md exists (the one place integrate may create a
  file), install the brain-report markers if absent, and compile so the block fills.
- `dsh`          — write the Agent Skill (the `.claude/skills` convention dsh shares),
  then PRINT the `cordis.patch.yml` insert that mounts brainpick through
  `@deepseek-ai/dsh-mcp-client` (dsh has no `claude mcp add`; a server is a config row).

The shipped Agent Skill (integrations/skill/SKILL.md, canonical) rides inside the
package; the parity test asserts the shipped copy is byte-identical to the canonical.
"""
from __future__ import annotations

import json
import os
import shlex
from pathlib import Path

from brainpick.compile.pipeline import run_compile
from brainpick.compile.t1 import REPORT_BEGIN_PREFIX, REPORT_END_MARKER
from brainpick.detect import find_repo_root
from brainpick.scaffold import _Voice, brainpick_command, dsh_snippet, mcp_snippets

TARGETS = ("claude-code", "opencode", "agents-md", "dsh")
HENXELS_BEGIN = "<!-- henxels:begin -->"

# harness -> where its Agent Skill lands, relative to the repo root
SKILL_DESTINATIONS = {
    "claude-code": Path(".claude") / "skills" / "brainpick" / "SKILL.md",
    "opencode": Path(".opencode") / "skills" / "brainpick" / "SKILL.md",
    # dsh loads Agent Skills from the same .claude/skills convention as Claude Code.
    "dsh": Path(".claude") / "skills" / "brainpick" / "SKILL.md",
}

_MINIMAL_AGENTS = "# AGENTS.md\n\nWorking notes for agents in this repository.\n"
_REPORT_PLACEHOLDER = (
    f"{REPORT_BEGIN_PREFIX}pending) -->\n"
    "_brainpick fills this block on the next `brainpick compile`._\n"
    f"{REPORT_END_MARKER}"
)


def skill_path() -> Path:
    """The shipped Agent Skill: the package copy first (installed wheels), then
    the repo-root canonical (dev checkout)."""
    packaged = Path(__file__).resolve().parent / "_skill" / "SKILL.md"
    if packaged.is_file():
        return packaged
    return Path(__file__).resolve().parents[4] / "integrations" / "skill" / "SKILL.md"


def skill_text() -> str:
    return skill_path().read_text(encoding="utf-8")


def _hooks_fragment(root: Path) -> str:
    """A paste-able Claude Code hooks fragment: a PreToolUse nudge toward the brain
    before it greps — advisory (exit 0), never a block — and the UserPromptSubmit
    recall that puts the matching memories into context (spec/72)."""
    recall = shlex.join(brainpick_command() + ["recall", "--root", str(root)])
    fragment = {
        "hooks": {
            "PreToolUse": [
                {
                    "matcher": "Grep|Glob",
                    "hooks": [
                        {
                            "type": "command",
                            "command": (
                                "echo 'brainpick: consult the brain first — brain_search "
                                "or `brainpick search` before grepping' >&2"
                            ),
                        }
                    ],
                }
            ],
            "UserPromptSubmit": [
                {"hooks": [{"type": "command", "command": recall, "timeout": 15}]}
            ],
        }
    }
    return json.dumps(fragment, indent=2)


def _write_skill(repo: Path, target: str, dry_run: bool) -> tuple[Path, bool]:
    dest = repo / SKILL_DESTINATIONS[target]
    existed = dest.is_file()
    if not dry_run:
        dest.parent.mkdir(parents=True, exist_ok=True)
        dest.write_text(skill_text(), encoding="utf-8")
    return dest, existed


def _insert_report_markers(text: str) -> str:
    """Install the report markers above the henxels digest block when one exists,
    else at the end — never disturbing hand-written content."""
    if REPORT_BEGIN_PREFIX in text:
        return text
    idx = text.find(HENXELS_BEGIN)
    if idx != -1:
        return text[:idx].rstrip("\n") + "\n\n" + _REPORT_PLACEHOLDER + "\n\n" + text[idx:]
    return text.rstrip("\n") + "\n\n" + _REPORT_PLACEHOLDER + "\n"


def _integrate_claude_code(voice: _Voice, root: Path, repo: Path, dry_run: bool) -> int:
    dest, existed = _write_skill(repo, "claude-code", dry_run)
    verb = "would write" if dry_run else ("updated" if existed else "wrote")
    voice.line("✓", f"skill: {verb} {dest}")
    if dry_run:
        voice.step("• print the graph-before-grep and recall hooks and the `claude mcp add` snippet")
        return 0
    voice.raw()
    voice.raw("Paste into .claude/settings.json (settings are never edited for you):")
    voice.raw()
    voice.raw(_hooks_fragment(root))
    voice.raw()
    voice.raw(mcp_snippets(root))
    return 0


def _integrate_opencode(voice: _Voice, root: Path, repo: Path, dry_run: bool) -> int:
    dest, existed = _write_skill(repo, "opencode", dry_run)
    verb = "would write" if dry_run else ("updated" if existed else "wrote")
    voice.line("✓", f"skill: {verb} {dest}")
    if dry_run:
        voice.step("• print the opencode.json MCP snippet")
        return 0
    voice.raw()
    voice.raw("Add the MCP server to opencode.json (merging JSON is left to you):")
    voice.raw()
    voice.raw(mcp_snippets(root))
    return 0


def _integrate_dsh(voice: _Voice, root: Path, repo: Path, dry_run: bool) -> int:
    dest, existed = _write_skill(repo, "dsh", dry_run)
    verb = "would write" if dry_run else ("updated" if existed else "wrote")
    voice.line("✓", f"skill: {verb} {dest}")
    if dry_run:
        voice.step("• print the cordis.patch.yml insert for @deepseek-ai/dsh-mcp-client")
        return 0
    voice.raw()
    voice.raw(dsh_snippet(root))
    return 0


def _integrate_agents_md(voice: _Voice, root: Path, repo: Path, dry_run: bool) -> int:
    agents = repo / "AGENTS.md"
    existed = agents.is_file()
    has_markers = existed and REPORT_BEGIN_PREFIX in agents.read_text(encoding="utf-8")

    if dry_run:
        if not existed:
            voice.step(f"• create a minimal {agents}")
        if not has_markers:
            voice.step(f"• install the brain-report markers in {agents}")
        voice.step(f"• compile {root} so the report block fills")
        return 0

    text = agents.read_text(encoding="utf-8") if existed else _MINIMAL_AGENTS
    if not existed:
        voice.line("✓", f"AGENTS.md: created {agents}")
    if REPORT_BEGIN_PREFIX not in text:
        text = _insert_report_markers(text)
        voice.line("✓", f"report: markers installed in {agents}")
    else:
        voice.line("○", f"report: markers already in {agents}")
    agents.write_text(text, encoding="utf-8")

    result = run_compile(root)
    voice.line("✓", f"compiled: the report block is filled (seq {result.seq})")
    voice.step(f"read it back: sed -n '/brainpick:begin report/,/brainpick:end report/p' {agents}")
    return 0


def run_integrate(target: str, root: str | Path, dry_run: bool = False) -> int:
    if target not in TARGETS:
        print(f"unknown target {target!r}; choose from {', '.join(TARGETS)}")
        return 1
    voice = _Voice(os.environ)
    voice.banner()
    root = Path(root).resolve()
    if not root.is_dir():
        voice.line("✗", f"{root} is not a directory")
        return 1
    repo = find_repo_root(root) or root
    voice.line("○", f"repo root: {repo}" + ("" if repo != root else " (the bundle is its own repo)"))
    if dry_run:
        voice.raw()
        voice.raw(f"dry run — nothing written. integrate {target} would:")

    if target == "claude-code":
        return _integrate_claude_code(voice, root, repo, dry_run)
    if target == "opencode":
        return _integrate_opencode(voice, root, repo, dry_run)
    if target == "dsh":
        return _integrate_dsh(voice, root, repo, dry_run)
    return _integrate_agents_md(voice, root, repo, dry_run)
